File: php/src/Algorithms/Graph.php
<?php

declare(strict_types=1);

namespace DSA\Algorithms;

/**
 * Graph representation using adjacency list.
 *
 * @template T
 */
class Graph
{
    /** @var array<int|string, array<int|string>> */
    private array $adjacencyList = [];
    private bool $directed;

    public function __construct(bool $directed = false)
    {
        $this->directed = $directed;
    }

    /**
     * Add a vertex to the graph.
     *
     * @param int|string $vertex
     */
    public function addVertex(int|string $vertex): void
    {
        if (!isset($this->adjacencyList[$vertex])) {
            $this->adjacencyList[$vertex] = [];
        }
    }

    /**
     * Add an edge between two vertices.
     *
     * @param int|string $from Source vertex
     * @param int|string $to Destination vertex
     */
    public function addEdge(int|string $from, int|string $to): void
    {
        $this->addVertex($from);
        $this->addVertex($to);

        if (!in_array($to, $this->adjacencyList[$from], true)) {
            $this->adjacencyList[$from][] = $to;
        }

        if (!$this->directed && !in_array($from, $this->adjacencyList[$to], true)) {
            $this->adjacencyList[$to][] = $from;
        }
    }

    /**
     * Remove an edge between two vertices.
     *
     * @param int|string $from Source vertex
     * @param int|string $to Destination vertex
     */
    public function removeEdge(int|string $from, int|string $to): void
    {
        if (isset($this->adjacencyList[$from])) {
            $this->adjacencyList[$from] = array_values(
                array_filter($this->adjacencyList[$from], fn($v) => $v !== $to)
            );
        }

        if (!$this->directed && isset($this->adjacencyList[$to])) {
            $this->adjacencyList[$to] = array_values(
                array_filter($this->adjacencyList[$to], fn($v) => $v !== $from)
            );
        }
    }

    /**
     * Remove a vertex and all its edges.
     *
     * @param int|string $vertex
     */
    public function removeVertex(int|string $vertex): void
    {
        if (!isset($this->adjacencyList[$vertex])) {
            return;
        }

        unset($this->adjacencyList[$vertex]);

        foreach ($this->adjacencyList as $v => $neighbors) {
            $this->adjacencyList[$v] = array_values(
                array_filter($neighbors, fn($n) => $n !== $vertex)
            );
        }
    }

    /**
     * Check if two vertices are connected by an edge.
     *
     * @param int|string $from Source vertex
     * @param int|string $to Destination vertex
     * @return bool
     */
    public function hasEdge(int|string $from, int|string $to): bool
    {
        return isset($this->adjacencyList[$from]) && in_array($to, $this->adjacencyList[$from], true);
    }

    /**
     * Get all neighbors of a vertex.
     *
     * @param int|string $vertex
     * @return array<int|string>
     */
    public function getNeighbors(int|string $vertex): array
    {
        return $this->adjacencyList[$vertex] ?? [];
    }

    /**
     * Get all vertices in the graph.
     *
     * @return array<int|string>
     */
    public function getVertices(): array
    {
        return array_keys($this->adjacencyList);
    }

    /**
     * Get the number of vertices.
     *
     * @return int
     */
    public function vertexCount(): int
    {
        return count($this->adjacencyList);
    }

    /**
     * Check if the graph is directed.
     *
     * @return bool
     */
    public function isDirected(): bool
    {
        return $this->directed;
    }

    /**
     * Get the adjacency list representation.
     *
     * @return array<int|string, array<int|string>>
     */
    public function getAdjacencyList(): array
    {
        return $this->adjacencyList;
    }
}

File: ruby/lib/algorithms/graph.rb
# frozen_string_literal: true

module DSA
  module Algorithms
    # Graph representation using adjacency list.
    class Graph
      attr_reader :directed

      def initialize(directed: false)
        @adjacency_list = {}
        @directed = directed
      end

      def add_vertex(vertex)
        @adjacency_list[vertex] ||= []
        self
      end

      def add_edge(from, to)
        add_vertex(from)
        add_vertex(to)

        @adjacency_list[from] << to unless @adjacency_list[from].include?(to)
        @adjacency_list[to] << from unless @directed || @adjacency_list[to].include?(from)
        self
      end

      def remove_edge(from, to)
        @adjacency_list[from]&.delete(to)
        @adjacency_list[to]&.delete(from) unless @directed
        self
      end

      def remove_vertex(vertex)
        @adjacency_list.delete(vertex)
        @adjacency_list.each_value { |neighbors| neighbors.delete(vertex) }
        self
      end

      def has_edge?(from, to)
        @adjacency_list[from]&.include?(to) || false
      end

      def neighbors(vertex)
        @adjacency_list[vertex]&.dup || []
      end

      def vertices
        @adjacency_list.keys
      end

      def vertex_count
        @adjacency_list.size
      end
    end
  end
end
